pkg/neotest/coverage.go

```diff
@@ -0,0 +1,189 @@
+package neotest
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"testing"
+
+	"github.com/nspcc-dev/neo-go/pkg/compiler"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+)
+
+var (
+	coverageLock sync.Mutex
+	// rawCoverage maps script hash to coverage data, collected during testing.
+	rawCoverage = make(map[util.Uint160]*scriptRawCoverage)
+	// flagChecked is true if `go test` coverage flag was checked at any point.
+	flagChecked bool
+	// coverageEnabled is true if coverage is being collected on test execution.
+	coverageEnabled bool
+	// coverProfile specifies the file all coverage data is written to, unless empty.
+	coverProfile = ""
+)
+
+type scriptRawCoverage struct {
+	debugInfo      *compiler.DebugInfo
+	offsetsVisited []int
+}
+
+type coverBlock struct {
+	// Line number for block start.
+	startLine uint
+	// Column number for block start.
+	startCol uint
+	// Line number for block end.
+	endLine uint
+	// Column number for block end.
+	endCol uint
+	// Number of statements included in this block.
+	stmts uint
+	// Number of times this block was executed.
+	counts uint
+}
+
+// documentName makes it clear when a `string` maps path to the script file
+type documentName = string
+
+func isCoverageEnabled() bool {
+	coverageLock.Lock()
+	defer coverageLock.Unlock()
+
+	if flagChecked {
+		return coverageEnabled
+	}
+	defer func() { flagChecked = true }()
+
+	const coverProfileFlag = "test.coverprofile"
+	flag.VisitAll(func(f *flag.Flag) {
+		if f.Name == coverProfileFlag && f.Value != nil {
+			coverageEnabled = true
+			coverProfile = f.Value.String()
+		}
+	})
+
+	if coverageEnabled {
+		// This is needed so go cover tool doesn't overwrite
+		// the file with our coverage when all tests are done.
+		err := flag.Set(coverProfileFlag, "")
+		if err != nil {
+			panic(err)
+		}
+	}
+	return coverageEnabled
+}
+
+var coverageHook vm.OnExecHook = func(scriptHash util.Uint160, offset int, opcode opcode.Opcode) {
+	coverageLock.Lock()
+	defer coverageLock.Unlock()
+	if cov, ok := rawCoverage[scriptHash]; ok {
+		cov.offsetsVisited = append(cov.offsetsVisited, offset)
+	}
+}
```

Comment:
I still think that it's a good idea to implement your optimisation:
>  so maybe we should leave coverageHook a function that accepts Executor in which OPs are collected and stored, then merged in reportCoverage

With the following adjustment:
> we still can't remove global rawCoverage variable because we need to merge the results of all tests and store them somewhere.

So I'd suggest:
1. Add `rawCoverage map[util.Uint160]*scriptRawCoverage` field to `Executor`.
2. Adjust `coverageHook` to be executor-specific and then collect OPs inside `Executor` instead of global variable.
3. Adjust `Cleanup` function of `(e *Executor) trackCoverage` so that firstly Executor's `rawCoverage` is merged into global `rawCoverage`, and only after that call `reportCoverage`.

We can make this optimisation in a separate PR, because the current PR is almost ready for the merge. @roman-khimov, agree?

Reply:
Ref. #3558.